diff --git a/common/src/main/java/me/chrr/camerapture/entity/ResizableDecorationEntity.java b/common/src/main/java/me/chrr/camerapture/entity/ResizableDecorationEntity.java
index b4b090fd8bd903dc6d874454977c8416294d20ab..b540b0c8c95a5be3171f0786e24cbd6a4e30432f 100644
--- a/common/src/main/java/me/chrr/camerapture/entity/ResizableDecorationEntity.java
+++ b/common/src/main/java/me/chrr/camerapture/entity/ResizableDecorationEntity.java
@@ -96,92 +96,106 @@ public abstract class ResizableDecorationEntity extends Entity {
     public int getFrameWidth() {
         return this.getDataTracker().get(FRAME_WIDTH);
     }
 
     public void setFrameWidth(int width) {
         this.getDataTracker().set(FRAME_WIDTH, width);
         updateBoundingBox();
     }
 
     public int getFrameHeight() {
         return this.getDataTracker().get(FRAME_HEIGHT);
     }
 
     public void setFrameHeight(int height) {
         this.getDataTracker().set(FRAME_HEIGHT, height);
         updateBoundingBox();
     }
 
     protected void updateBoundingBox() {
         if (this.facing == null) {
             return;
         }
 
         // Find the center of the top-left piece.
         Vec3d center = attachmentPos.toCenterPos();
-        center = center.subtract(new Vec3d(facing.getOffsetX(), 0, facing.getOffsetZ()).multiply(0.5 - THICKNESS / 2));
+        center = center.subtract(new Vec3d(facing.getOffsetX(), facing.getOffsetY(), facing.getOffsetZ()).multiply(0.5 - THICKNESS / 2));
 
         this.setPos(center.x, center.y, center.z);
 
         // Then, we expand into the full frame.
-        Direction parallel = facing.rotateYCounterclockwise();
-        if (facing.getAxis() == Direction.Axis.Z) {
-            Vec3d p1 = center.subtract(parallel.getOffsetX() * 0.5, 0.5, THICKNESS / 2);
-            Vec3d p2 = p1.add(parallel.getOffsetX() * getFrameWidth(), getFrameHeight(), THICKNESS);
+        if (facing.getAxis() == Direction.Axis.Y) {
+            Vec3d p1 = center.subtract(0.5, THICKNESS / 2, 0.5);
+            Vec3d p2 = p1.add(getFrameWidth(), THICKNESS, getFrameHeight());
             this.setBoundingBox(new Box(p1, p2));
         } else {
-            Vec3d p1 = center.subtract(THICKNESS / 2, 0.5, parallel.getOffsetZ() * 0.5);
-            Vec3d p2 = p1.add(THICKNESS, getFrameHeight(), parallel.getOffsetZ() * getFrameWidth());
-            this.setBoundingBox(new Box(p1, p2));
+            Direction parallel = facing.rotateYCounterclockwise();
+            if (facing.getAxis() == Direction.Axis.Z) {
+                Vec3d p1 = center.subtract(parallel.getOffsetX() * 0.5, 0.5, THICKNESS / 2);
+                Vec3d p2 = p1.add(parallel.getOffsetX() * getFrameWidth(), getFrameHeight(), THICKNESS);
+                this.setBoundingBox(new Box(p1, p2));
+            } else {
+                Vec3d p1 = center.subtract(THICKNESS / 2, 0.5, parallel.getOffsetZ() * 0.5);
+                Vec3d p2 = p1.add(THICKNESS, getFrameHeight(), parallel.getOffsetZ() * getFrameWidth());
+                this.setBoundingBox(new Box(p1, p2));
+            }
         }
 
         resetObstructionCheckCounter();
     }
 
     @Override
     public Direction getHorizontalFacing() {
         return getFacing();
     }
 
     @Override
     public boolean canHit() {
         return true;
     }
 
     public boolean canStayAttached() {
         if (!Camerapture.CONFIG_MANAGER.getConfig().server.checkFramePosition) {
             return true;
         } else if (!this.getWorld().isSpaceEmpty(this)) {
             return false;
         } else {
             BlockPos blockPos = this.attachmentPos.offset(this.facing.getOpposite());
-            Direction direction = this.facing.rotateYCounterclockwise();
+            Direction widthDir;
+            Direction heightDir;
+            if (this.facing.getAxis() == Direction.Axis.Y) {
+                widthDir = Direction.EAST;
+                heightDir = Direction.SOUTH;
+            } else {
+                widthDir = this.facing.rotateYCounterclockwise();
+                heightDir = Direction.UP;
+            }
             BlockPos.Mutable mutable = new BlockPos.Mutable();
 
             for (int x = 0; x < this.getFrameWidth(); ++x) {
                 for (int y = 0; y < this.getFrameHeight(); ++y) {
-                    mutable.set(blockPos).move(direction, x).move(Direction.UP, y);
+                    mutable.set(blockPos).move(widthDir, x).move(heightDir, y);
                     BlockState blockState = this.getWorld().getBlockState(mutable);
 
                     //noinspection deprecation
                     if (!blockState.isSolid() && !AbstractRedstoneGateBlock.isRedstoneGate(blockState)) {
                         return false;
                     }
                 }
             }
 
             return this.getWorld()
                     .getOtherEntities(this, this.getBoundingBox(), (entity) ->
                             entity instanceof AbstractDecorationEntity || entity instanceof ResizableDecorationEntity)
                     .isEmpty();
         }
     }
 
     @Override
     public boolean handleAttack(Entity attacker) {
         if (attacker instanceof PlayerEntity playerEntity) {
             //noinspection deprecation: let's just copy what Vanilla does for now.
             return !this.getWorld().canEntityModifyAt(playerEntity, this.getBlockPos())
                     || this.sidedDamage(this.getDamageSources().playerAttack(playerEntity), 0.0F);
         } else {
             return false;
         }
diff --git a/common/src/main/java/me/chrr/camerapture/item/PictureItem.java b/common/src/main/java/me/chrr/camerapture/item/PictureItem.java
index 6f457b45b14ff42ab3081e41d65cb1793058848c..4d68e23c31b58e35ec45cb9056449a8fe4a557d7 100644
--- a/common/src/main/java/me/chrr/camerapture/item/PictureItem.java
+++ b/common/src/main/java/me/chrr/camerapture/item/PictureItem.java
@@ -42,56 +42,57 @@ public class PictureItem extends Item {
         super(new Settings().registryKey(KEY));
     }
 
     @Override
     public ActionResult use(World world, PlayerEntity user, Hand hand) {
         if (!user.isSneaking()) {
             // When not sneaking, we show the picture client-side through an event handler.
             return ActionResult.SUCCESS;
         } else {
             return ActionResult.PASS;
         }
     }
 
     @Override
     public ActionResult useOnBlock(ItemUsageContext context) {
         PlayerEntity player = context.getPlayer();
         if (player == null || !player.isSneaking()) {
             return ActionResult.PASS;
         }
 
         World world = context.getWorld();
         Direction facing = context.getSide();
         BlockPos pos = context.getBlockPos().offset(facing);
         ItemStack itemStack = context.getStack();
 
-        // Pictures can only be placed on walls.
-        if (facing.getAxis().isVertical() || !player.canPlaceOn(pos, facing, itemStack)) {
+        // Pictures must be placeable on the clicked surface
+        if (!player.canPlaceOn(pos, facing, itemStack)) {
             return ActionResult.PASS;
         }
 
         PictureFrameEntity pictureFrameEntity = new PictureFrameEntity(world, pos, facing);
+        pictureFrameEntity.setYaw(player.getYaw());
         if (!pictureFrameEntity.canStayAttached()) {
             return ActionResult.PASS;
         }
 
         // Correctly handle (+NBT) items
         NbtComponent nbtComponent = itemStack.getOrDefault(DataComponentTypes.ENTITY_DATA, NbtComponent.DEFAULT);
         if (!nbtComponent.isEmpty()) {
             EntityType.loadFromEntityNbt(world, player, pictureFrameEntity, nbtComponent);
         }
 
         pictureFrameEntity.setItemStack(itemStack.copyWithCount(1));
 
         if (!world.isClient) {
             pictureFrameEntity.onPlace();
             world.emitGameEvent(player, GameEvent.ENTITY_PLACE, pictureFrameEntity.getPos());
             world.spawnEntity(pictureFrameEntity);
         }
 
         itemStack.decrement(1);
         return ActionResult.SUCCESS;
     }
 
     public static ItemStack create(String creator, UUID uuid) {
         ItemStack stack = new ItemStack(Camerapture.PICTURE, 1);
         stack.set(Camerapture.PICTURE_DATA, new PictureData(uuid, creator, System.currentTimeMillis()));
